File: parsers/parser.go
package parsers

import (
	s "apivapt/schema"
	"net/http"
)

type Parser interface {
	Parse(data []byte) (*s.APISchema, error)
	Detect(baseURL string, header http.Header, body []byte) (bool, string)
	HasRoutes(data []byte) bool
}

type WordpressParser struct{}
type OpenAPIParser struct{}
type GraphQLParser struct{}

File: parser/parser.py
import httpx

from .base import BaseParser
from .wordpress import WordpressParser
from .openapi import OpenAPIParser
from .graphql import GraphQLParser


class Parser:
    parsers: list[BaseParser] = [WordpressParser(), OpenAPIParser(), GraphQLParser()]

    def Detect(
        self, base_url: str, header: httpx.Headers, body: str
    ) -> tuple[BaseParser | None, str]:
        for p in self.parsers:
            ok, doc_url = p.Detect(base_url, header, body)
            if ok:
                return p, doc_url
        return None, ""
